libraries/Microsoft.Bot.Builder.Azure/AzureTableStorage.cs

```diff
@@ -10,157 +10,269 @@
 
-            public StoreItemEntity(EntityKey key, object obj)
-                : this(key.PartitionKey, key.RowKey, obj)
-            { }
+            async Task<CloudTable> EnsureTableExists()
+            {
+                _table = _storageAccount.CreateCloudTableClient().GetTableReference(_tableName);
+                
+                // This call may not be thread-safe and multiple calls may be made, but Azure will handle it correctly and there is no destructive side effect.
+                await _table.CreateIfNotExistsAsync();
+                return _table;
+            }
+        }
 
-            public StoreItemEntity(string partitionKey, string rowKey, object obj)
-                : base(partitionKey, rowKey)
+        /// <summary>
+        /// Internal data structure for storing items in Azure Tables.
+        /// </summary>
+        private class StorageEnvelope
+        {
+            public object StoreItem { get; private set; }
+            public EntityKey Key { get; private set; }
+            public string ETag
             {
-                this.ETag = (obj as IStoreItem)?.eTag;
-                this.Json = JsonConvert.SerializeObject(obj, Formatting.None, serializationSettings);
+                get
+                {
+                    return (StoreItem as IStoreItem)?.eTag;
+                }
             }
 
-            public string Json { get; set; }
+            public StorageEnvelope(EntityKey key, object entity)
+            {
+                Key = key;
+                StoreItem = entity;
+            }
 
-            public object AsObject()
+            public DynamicTableEntity AsTableEntity()
             {
-                var obj = JsonConvert.DeserializeObject(Json, serializationSettings);
-                IStoreItem storeItem = obj as IStoreItem;
-                if (storeItem != null)
-                    storeItem.eTag = this.ETag;
-                return obj;
+                // Flatten properties
+                var properties = EntityPropertyConverter.Flatten(StoreItem, new OperationContext());
+
+                // Add Type information
+                var type = StoreItem.GetType();
+                var typeQualifiedName = type.AssemblyQualifiedName;
+                properties.Add("__type", EntityProperty.GeneratePropertyForString(typeQualifiedName));
+
+                return new DynamicTableEntity(Key.PartitionKey, Key.RowKey)
+                {
+                    ETag = ETag,
+                    Properties = properties
+                };
+            }
+
+            public static StorageEnvelope AsStoreItemEntity(TableResult tableEntity)
+            {
+                // Create instance of proper type
+                var dynamicTableEntity = (DynamicTableEntity)tableEntity.Result;
+                var type = Type.GetType(dynamicTableEntity.Properties["__type"].StringValue);
+                var properties = dynamicTableEntity.Properties;
```

Comment:
Consider switching the order of these two lines and using the local variable for properties instead of calling the property twice:

```CSharp
var properties = dynamicTableEntity.Properties; 
var type = Type.GetType(properties["__type"].StringValue); 
```